ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,241 @@
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Exposes stats related to a {@link BlobStore} that is useful to different components.
+ *
+ * Note: This is the v0 implementation of BlobStoreStats. The v0 implementation walks through the entire index
+ * and collect data needed to serve stats related requests.
+ */
+public class BlobStoreStats implements StoreStats {
+  static final String IO_SCHEDULER_JOB_TYPE = "BlobStoreStats";
+  static final String IO_SCHEDULER_JOB_ID = "indexSegment_read";
+
+  private final PersistentIndex index;
+  private final Time time;
+  private final DiskIOScheduler diskIOScheduler;
+
+  BlobStoreStats(PersistentIndex index, Time time, DiskIOScheduler diskIOScheduler) {
+    this.index = index;
+    this.time = time;
+    this.diskIOScheduler = diskIOScheduler;
+  }
+
+  // TODO remove this once #541 is merged
+  public static String getServiceId(StoreKey key) {
+    return key.getID().substring(0, 1);
+  }
+
+  // TODO remove this once #541 is merged
+  public static String getContainerId(StoreKey key) {
+    return key.getID().substring(key.getID().length() - 1);
+  }
+
+  @Override
+  public Pair<Long, Long> getValidDataSize(TimeRange timeRange) throws StoreException {
+    Pair<Long, Map<String, Long>> logSegmentValidSizeResult = getValidDataSizeByLogSegment(timeRange);
+    Long allLogSegmentValidDataSize = 0L;
+    for (Long value : logSegmentValidSizeResult.getSecond().values()) {
+      allLogSegmentValidDataSize += value;
+    }
+    return new Pair<>(logSegmentValidSizeResult.getFirst(), allLogSegmentValidDataSize);
+  }
+
+  @Override
+  public Pair<Long, Map<String, Long>> getValidDataSizeByLogSegment(TimeRange timeRange) throws StoreException {
+    long referenceTimeInSecs = timeRange.getEndTimeInSecs();
+    Map<String, Long> logSegmentValidSizeMap = collectValidDataSizeByLogSegment(referenceTimeInSecs * Time.MsPerSec);
+    return new Pair<>(referenceTimeInSecs, logSegmentValidSizeMap);
+  }
+
+  @Override
+  public Map<String, Map<String, Long>> getValidDataSize() throws StoreException {
+    long referenceTimeInMs = time.milliseconds();
+    return collectValidDataSizeByContainer(referenceTimeInMs);
+  }
+
+  /**
+   * Walk through the entire index and collect valid data size information per container.
+   * @param referenceTimeInMs the reference time to be used to decide whether or not a blob is valid
+   * @return a nested {@link Map} of serviceId to containerId to valid data size
+   */
+  private Map<String, Map<String, Long>> collectValidDataSizeByContainer(long referenceTimeInMs)
+      throws StoreException {
+    Offset previousSegmentEndOffset = null;
+    Offset indexStartOffset = index.getStartOffset();
+    Map<String, Map<String, Long>> containerValidSizeMap = new HashMap<>();
+    for (IndexSegment indexSegment : index.getIndexSegments().values()) {
+      diskIOScheduler.getSlice(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, BlobStoreStats.IO_SCHEDULER_JOB_ID, 1);
```

Comment:
we actually call the throttler after we have performed some operation. Look at it this way. After performing N units of work, we inform the throttler that we have completed N units of work. In the same call, the throttle will wait if it feels we are above the set threshold limits. So, it makes sense to call the throttler once we have done some unit of work. Doing it before does not cause any harm though, just that this is more understandable.